android/app/src/main/java/io/github/parkwithease/parkeasy/ui/login/LoginViewModel.kt

```diff
@@ -30,40 +37,144 @@ constructor(
+            _formEnabled.value = true
+        }
+    }
+
+    fun onNameChange(value: String) {
+        formState = formState.run { copy(name = name.copy(value = value)) }
+    }
+
+    fun onEmailChange(value: String) {
+        formState = formState.run { copy(email = email.copy(value = value)) }
+    }
+
+    fun onPasswordChange(value: String) {
+        formState =
+            formState.run {
+                copy(
+                    password = password.copy(value = value),
+                    confirmPassword =
+                        confirmPassword.copy(
+                            error =
+                                if (value == confirmPassword.value) null
+                                else "Password does not match"
+                        ),
+                )
+            }
+    }
+
+    fun onConfirmPasswordChange(value: String) {
+        formState =
+            formState.run {
+                copy(
+                    confirmPassword =
+                        confirmPassword.copy(
+                            value = value,
+                            error = if (value == password.value) null else "Password does not match",
+                        )
+                )
+            }
+    }
+
+    private fun Result<Unit>.recoverRequestErrors(operationFailMsg: String): Result<Unit> =
+        recover {
+            when (it) {
+                is APIException -> {
+                    errorToForm(it.error)
+                    viewModelScope.launch { showSnackbar(operationFailMsg, null) }
+                }
+                is IOException -> {
+                    viewModelScope.launch {
+                        showSnackbar("Could not connect to server, are you online?", null)
+                    }
+                }
+                else -> throw it
+            }
+        }
+
+    private fun errorToForm(error: ErrorModel) {
+        when (error.type) {
+            ErrorModel.TYPE_INVALID_CREDENTIALS -> {
+                formState =
+                    formState.run {
+                        copy(email = email.copy(error = ""), password = password.copy(error = ""))
+                    }
+            }
+
+            ErrorModel.TYPE_PASSWORD_LENGTH ->
+                formState =
+                    formState.run {
+                        copy(password = password.copy(error = "Password too long or too short"))
+                    }
+
+            else -> annotateErrorLocation(error.errors)
+        }
+    }
+
+    private fun annotateErrorLocation(errors: List<ErrorDetail>) {
+        for (err in errors) {
+            when (err.location) {
+                "body.email" ->
+                    formState =
+                        formState.run { copy(email = email.copy(error = "Invalid email address")) }
+
+                "body.password" ->
+                    formState =
+                        formState.run { copy(password = password.copy(error = "Invalid password")) }
+            }
         }
```

Comment:
Errors don't properly get cleared which leads to misleading error messages

Reply:
Having a password error makes the email error persist
![image](https://github.com/user-attachments/assets/d31dcda0-c74e-4b3f-a57b-63ce684bcb6c)
![image](https://github.com/user-attachments/assets/b759ee06-1d26-47a9-a1c2-8fde7b2de12a)